firmware/pattern.c

```diff
@@ -26,7 +27,9 @@ static const unsigned int color_bar[8] = {
 	YCBCR422_BLUE,
 	YCBCR422_BLACK
 };
+#endif
 
+#ifdef MAIN_RAM_BASE
```

Comment:
I'm not sure why this define is here? Being `static const` should mean it doesn't end up in the SRAM?

Reply:
Without this gate (and without `FIRMWARE=stub`), I get the following error:

```
 CC       pattern.o
C:/msys64/home/William/prj/Consult/timvideos/litex-buildenv/firmware/pattern.c:161:13: warning: 'pattern_draw_text' defined but not used [-Wunused-function]
 static void pattern_draw_text(int x, int y, char *ptr) {
             ^~~~~~~~~~~~~~~~~
C:/msys64/home/William/prj/Consult/timvideos/litex-buildenv/firmware/pattern.c:133:12: warning: 'inc_color' defined but not used [-Wunused-function]
 static int inc_color(int color) {
            ^~~~~~~~~
C:/msys64/home/William/prj/Consult/timvideos/litex-buildenv/firmware/pattern.c:33:28: error: 'font5x7' defined but not used [-Werror=unused-const-variable=]
 static const unsigned char font5x7[] = {
                            ^~~~~~~
C:/msys64/home/William/prj/Consult/timvideos/litex-buildenv/firmware/pattern.c:20:27: error: 'color_bar' defined but not used [-Werror=unused-const-variable=]
 static const unsigned int color_bar[8] = {
                           ^~~~~~~~~
cc1.exe: all warnings being treated as errors
make[1]: *** [C:/msys64/home/William/prj/Consult/timvideos/litex-buildenv/firmware/Makefile:106: pattern.o] Error 1
make[1]: Leaving directory '/home/William/prj/Consult/timvideos/litex-buildenv/build/tinyfpga_bx_base_lm32.lite/software/firmware'
Traceback (most recent call last):
  File "./make.py", line 151, in <module>
    main()
  File "./make.py", line 135, in main
    vns = builder.build(**dict(args.build_option))
  File "C:/msys64/home/william/prj/projects/litex\litex\soc\integration\builder.py", line 160, in build
    self._generate_software(not self.soc.integrated_rom_initialized)
  File "C:/msys64/home/william/prj/projects/litex\litex\soc\integration\builder.py", line 134, in _generate_software
    subprocess.check_call(["make", "-C", dst_dir, "-f", makefile])
  File "C:/msys64/mingw64/lib/python3.6\subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['make', '-C', 'C:/msys64/home/William/prj/Consult/timvideos/litex-buildenv/build/tinyfpga_bx_base_lm32.lite/software/firmware', '-f', 'C:/msys64/home/William/prj/Consult/timvideos/litex-buildenv/firmware/Makefile']' returned non-zero exit status 2.
```

I added the gate, based on the following [comment](https://github.com/timvideos/litex-buildenv/blob/3f691e817cb7be802b760646ff84a572ee8e28d9/firmware/pattern.c#L157):

```
/* FIXME: Framebuffer Should not even be compiled if no MAIN RAM */
```

So it's a band-aid :P, but what do you think we should do?

Reply:
@cr1901 @mithro It looks like the *correct* fix would be to have a BIOS build target that does not even build/link `pattern.c`.  AFAICT that's something that really only makes sense on video output boards (Opsis, etc).  Eg, maybe it shouldn't be built for `TARGET=BASE`?

Ewen

Reply:
Looking into this deeper, it seems that in C, these values are not optimized away (compared to C++ were they would be). From https://nullprogram.com/blog/2016/07/25/;
> In general, const declarations can’t help the optimizer, but const definitions can.

Reply:
FWIW, I've found my commit (from pre-LCA2018 hack fest) which I think included the `FIXME`, etc) when we were trying to get the BIOS to fit into a smaller system (that one ultimately didn't work because the Spartan 3 FPGA part was *tiny* and not even a minimal CPU would fit).   

Looks like [that change eventually got merged some other way](https://github.com/timvideos/litex-buildenv/commit/87b46c0248d82cc1211464ba582da196e802bb1b) (maybe via someone else trying to get things working).  But [comment on commit explains the rationale for the first hack](  https://github.com/ewenmcneill/litex-buildenv/commit/87b46c0248d82cc1211464ba582da196e802bb1b), which I hadn't intended to be merged in that form...  (I've still got the branch sitting around as I had assumed it didn't get merged!)

Ewen

Reply:
`TARGET` is available to the `Makefile`. Since by convention, `base` systems don't have video, it's probably alright to modify the firmware `Makefile` to only append `pattern.c` if we are compiling a design more comprehensive than `TARGET=base`.

Reply:
@cr1901 @mithro I've created https://github.com/timvideos/litex-buildenv/issues/59 to track this desire to build a minimised version of the BIOS for `TARGET=base`.  I remember us talking about it in January too, but it looks like the kludgey "`#ifdef` some things out to find out if it'll even fit" version got merged as-is (presumably along with some other changes).  I'd suggest taking this pull request including the extra kludges in `firmware/pattern.c` and working on `Makefile` changes for a more minimal BIOS separately under https://github.com/timvideos/litex-buildenv/issues/59.  

Ewen